gpii/node_modules/preferencesServer/src/prefsSafesHandlers.js

```diff
@@ -0,0 +1,129 @@
+            args: ["{gpii.preferencesServer.preferencesService}", "{preferencesServer}", "{that}", "{that}.req.query.view"]
+        }
+    }
+});
+
+gpii.preferencesServer.prefsSafeList.handler.getPrefsSafeList = function (prefsService, preferencesServer, request, view) {
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listPrefsSafesDataSource.options.directModel, {});
+    var prefsPromise = prefsService.dataStore.listPrefsSafesDataSource.get(finalDirectModel);
+    prefsPromise.then(function (data) {
+        var togo = [];
+        // TODO this should probably be done in a couchview for efficiency and eventual paging?
+        fluid.each(data, function(item) {
+            item.value.prefsSafeId = item.key;
+            togo.push(item.value);
+        });
+        request.events.onSuccess.fire(togo);
+    }, request.events.onError.fire);
+};
+
+fluid.defaults("gpii.preferencesServer.listKeysForPrefsSafe.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.preferencesServer.listKeysForPrefsSafe.handler.getKeysForPrefsSafe",
+            args: ["{gpii.preferencesServer.preferencesService}", "{preferencesServer}", "{that}", "{that}.req.params.prefsSafeId", "{that}.req.query.view"]
+        }
+    }
+});
+
+gpii.preferencesServer.listKeysForPrefsSafe.handler.getKeysForPrefsSafe = function (prefsService, preferencesServer, request, prefsSafeId, view) {
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel, {
+        prefsSafeId: prefsSafeId});
+    var prefsPromise = prefsService.dataStore.listKeysForPrefsSafeDataSource.get(finalDirectModel);
+    prefsPromise.then(function (data) {
+        var togo = [];
+        fluid.each(data, function (i) { 
+            var next = i.doc;
+            gpii.dbOperation.dbDataStore.cleanUpDoc(next);
+            togo.push(next);
+        });
+        request.events.onSuccess.fire(togo);
+    }, request.events.onError.fire);
+};
+
+fluid.defaults("gpii.preferencesServer.prefsSafeWithKeysGet.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.preferencesServer.prefsSafeWithKeysGet.handler.get",
+            args: ["{gpii.preferencesServer.preferencesService}", "{preferencesServer}", "{that}", "{that}.req.params.prefsSafeId", "{that}.req.query.view"]
+        }
+    }
+});
+
+gpii.preferencesServer.prefsSafeWithKeysGet.handler.get = function (prefsService, preferencesServer, request, prefsSafeId, view) {
+    // TODO Refactor from above, copied/pasted
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel, {
+        prefsSafeId: prefsSafeId});
+    var prefsPromise = prefsService.dataStore.listKeysForPrefsSafeDataSource.get(finalDirectModel);
```

Comment:
All handlers in this script are accessing `prefsService.dataStore` directly, which is not recommended. The `prefsService` acts as a bridge that helps to decouple the database operation from other GPII modules including preferencesServer. `prefsService` should provide functions that make sense to request handlers instead of having request handlers directly deal with data that are in raw document structure. I would suggest to move the 2 steps handling in this function to `prefsService`.

This comment applies to other handlers in this script.